lib/ASRouterTriggerListeners.jsm

```diff
@@ -8,11 +8,138 @@ const {Services} = ChromeUtils.import("resource://gre/modules/Services.jsm");
 ChromeUtils.defineModuleGetter(this, "PrivateBrowsingUtils",
   "resource://gre/modules/PrivateBrowsingUtils.jsm");
 
+const FEW_MINUTES = 15 * 60 * 1000; // 15 mins
+
+/**
+ * Wait for browser startup to finish to avoid accessing uninitialized
+ * properties
+ */
+async function checkStartupFinished(win) {
+  if (!win.gBrowserInit.delayedStartupFinished) {
+    await new Promise(resolve => {
+      let delayedStartupObserver = (subject, topic) => {
+        if (topic === "browser-delayed-startup-finished" && subject === win) {
+          Services.obs.removeObserver(delayedStartupObserver, "browser-delayed-startup-finished");
+          resolve();
+        }
+      };
+
+      Services.obs.addObserver(delayedStartupObserver, "browser-delayed-startup-finished");
+    });
+  }
+}
+
 /**
  * A Map from trigger IDs to singleton trigger listeners. Each listener must
  * have idempotent `init` and `uninit` methods.
  */
 this.ASRouterTriggerListeners = new Map([
+  ["frequentVisits", {
+    _initialized: false,
+    _triggerHandler: null,
+    _hosts: null,
+    _visits: null,
+
+    async init(triggerHandler, hosts) {
+      if (this._initialized) {
+        return;
+      }
+      this.onTabSwitch = this.onTabSwitch.bind(this);
+
+      // Add listeners to all existing browser windows
+      for (let win of Services.wm.getEnumerator("navigator:browser")) {
+        if (win.closed || PrivateBrowsingUtils.isWindowPrivate(win)) {
+          continue;
+        }
+        await checkStartupFinished(win);
```

Comment:
Do you need to do this before checking `PrivateBrowsingUtils.isWindowPrivate` or does that not matter?

Reply:
That's meant to guard against `win.gBrowser` being undefined.

Reply:
Yeah I mean is `PrivateBrowsingUtils.isWindowPrivate` going to throw if `win.gBrowser` is undefined (i.e. should we move that part after this check)

Reply:
Sorry I didn't give full context. That line was added because the access to `win.gBrowser` threw an error in mochitests. Checking to see if it is a private window never caused any issues.